src/Dialect/ONNX/ONNXOps/Canonicalize.cpp

```diff
@@ -1239,6 +1239,58 @@ class PowToMulRewritePattern : public OpRewritePattern<ONNXPowOp> {
   int64_t maxPower;
 };
 
+class PowConversionRewritePattern : public OpRewritePattern<ONNXPowOp> {
+public:
+  using OpRewritePattern<ONNXPowOp>::OpRewritePattern;
+
+  PowConversionRewritePattern(MLIRContext *context, int64_t maxPower)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(
+      ONNXPowOp powOp, PatternRewriter &rewriter) const override {
+    Operation *op = powOp.getOperation();
+    Location loc = powOp.getLoc();
+
+    Value input = powOp.getX();
+    Value exp = powOp.getY();
+    if (!hasShapeAndRank(input) || !hasShapeAndRank(exp))
+      return failure(); // Cannot apply pattern until ranks are known.
+
+    // Characterize element types.
+    Type inputElementType =
+        mlir::cast<ShapedType>(input.getType()).getElementType();
+    Type expElementType =
+        mlir::cast<ShapedType>(exp.getType()).getElementType();
+    bool inputIsInt = isa<IntegerType>(inputElementType);
+    bool inputIsFloat = isa<FloatType>(inputElementType);
+    bool expIsInt = isa<IntegerType>(expElementType);
+    bool expIsFloat = isa<FloatType>(expElementType);
+    int64_t inputWidth = 0, expWidth = 0;
```

Comment:
this can be simplified with `getElementTypeBitWidth` from a `ShapedType` or `getIntOrFloatBitWidth` from a `Type` that's expected to be `FloatType` or `IntegerType`

Reply:
good suggestion, done.